aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -0,0 +1,804 @@
+#include <THC/THCThrustAllocator.cuh>
+
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAUtils.h>
+
+#include <ATen/native/sparse/cuda/SparseCUDABlas.cuh>
+#include <c10/cuda/CUDACachingAllocator.h>
+
+#include <thrust/device_vector.h>
+#include <thrust/host_vector.h>
+
+#if defined(__CUDACC__) && (CUSPARSE_VERSION >= 11000)
+#define IS_CUSPARSE11_AVAILABLE() 1
+#else
+#define IS_CUSPARSE11_AVAILABLE() 0
+#endif
+
+#if IS_CUSPARSE11_AVAILABLE()
+#include <library_types.h>
+#endif
+
+namespace at {
+namespace native {
+
+namespace {
+
+using namespace at::sparse;
+
+IntTensor _to_csr_int(const LongTensor& rowIndices, int64_t dim, int64_t nnz) {
+  IntTensor csr = at::empty({dim + 1}, CUDA(kInt));
+  IntTensor rowIndicesInt = at::empty({rowIndices.size(0)}, CUDA(kInt));
+  rowIndicesInt.copy_(rowIndices);
+  sparse::cuda::Xcoo2csr(
+      rowIndicesInt.data_ptr<int32_t>(), nnz, dim, csr.data_ptr<int32_t>());
+  return csr;
+}
+
+int confirm_mult_size(const std::vector<int>& mat1_size, const std::vector<int>& mat2_size) {
+  TORCH_CHECK(
+      mat1_size[1] == mat2_size[0],
+      "mat1 and mat2 shapes cannot be multiplied (",
+      mat1_size[0],
+      "x",
+      mat1_size[1],
+      " and ",
+      mat2_size[0],
+      "x",
+      mat2_size[1],
+      ")");
+  return mat1_size[1];
+}
+
+void create_general_description_(cusparseMatDescr_t& description_) {
+  TORCH_CUDASPARSE_CHECK(cusparseCreateMatDescr(&description_));
+  TORCH_CUDASPARSE_CHECK(cusparseSetMatType(description_, CUSPARSE_MATRIX_TYPE_GENERAL));
+  TORCH_CUDASPARSE_CHECK(cusparseSetMatIndexBase(description_, CUSPARSE_INDEX_BASE_ZERO));
+}
+
+struct csrOutput {
+  IntTensor csr_indices_{};
+  IntTensor csr_pointers_{};
+  at::Tensor csr_values_{};
+  int nnz_{0};
+  std::vector<int> size_;
+
+  cusparseMatDescr_t description_{0};
+
+  csrOutput(const std::vector<int> &size) : size_{size} {
+    create_general_description_(description_);
+  }
+
+  ~csrOutput() {
+    cusparseDestroyMatDescr(description_);
+  }
+
+  int size(int index) const {
+    return size_.at(index);
+  }
+};
+
+template<class scalar_t> 
+struct csrMatrixRef {
+  int* csr_indices_{nullptr};
+  int* csr_pointers_{nullptr};
+  scalar_t* csr_values_{nullptr};
+  int nnz_{0};
+  std::vector<int> size_{};
+
+  #if IS_CUSPARSE11_AVAILABLE()
+    cusparseSpMatDescr_t description_{0};
+  #else 
+    cusparseMatDescr_t description_{0};
+  #endif
+
+  csrMatrixRef() {
+    #if !IS_CUSPARSE11_AVAILABLE()
+      create_general_description_(description_);
+    #endif
+  }
+
+  csrMatrixRef(
+      int* csr_indices,
+      int* csr_pointers,
+      scalar_t* csr_values,
+      int nnz,
+      const std::vector<int>& size)
+      : csr_indices_{csr_indices},
+        csr_pointers_{csr_pointers},
+        csr_values_{csr_values},
+        nnz_{nnz},
+        size_{size} {
+    #if IS_CUSPARSE11_AVAILABLE()
+      cudaDataType cuda_data_type;
+      if constexpr ( std::is_same<float, scalar_t>::value ) {
```

Comment:
Does `if constexpr` work? I thought it wasn't available until C++17, and PyTorch supports C++14. Note we also have this:

https://github.com/pytorch/pytorch/blob/8177f63c91f7c6343859d12c6818e129acaa974d/c10/util/C%2B%2B17.h#L189

Which I thought was intended to let us use a `if constexpr` like construct without having to be compiled with C++17 support.

Reply:
@ngimel suggests just removing the `if constexpr` construct for simplicity